osf/management/commands/purge_files.py

```diff
@@ -0,0 +1,98 @@
+import logging
+
+from django.utils import timezone
+
+from framework.celery_tasks import app as celery_app
+from website.app import setup_django
+setup_django()
+from osf.models import Registration
+from addons.osfstorage.models import TrashedFileNode
+from django.core.management.base import BaseCommand
+from osf.utils.workflows import RegistrationModerationStates
+from website.settings import GCS_CREDS, PURGE_DELTA
+from google.cloud.storage.client import Client
+from google.oauth2.service_account import Credentials
+
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def mark_withdrawn_files_as_deleted(dry_run=False):
+    for node in Registration.objects.filter(moderation_state=RegistrationModerationStates.WITHDRAWN.db_name):
+        for file in node.files.all():
+            if not dry_run:
+                file.delete()
+
+
+def purge_deleted_withdrawn_files(dry_run=False):
+    deleted_file_ids_on_withdrawn_node = Registration.objects.filter(
+        moderation_state=RegistrationModerationStates.WITHDRAWN.db_name,
+        files__deleted__lte=timezone.now() - PURGE_DELTA
+    ).values_list('files__id', flat=True).distinct()
+
+    creds = Credentials.from_service_account_file(GCS_CREDS)
+    client = Client(credentials=creds)
+    for file in TrashedFileNode.objects.filter(id__in=deleted_file_ids_on_withdrawn_node):
+        if not dry_run:
+            file._purge(client=client)
+
+
+def mark_failed_registration_files_as_deleted(dry_run=False):
+    """
+    These registrations have ArchiveJobs stuck in initial state after the archive timeout period. Since they have timed
+    out they can be marked as deleted so they can be purged in PURGE_DELTA days.
+    """
+    for reg in Registration.find_failed_registrations():
+        reg.delete_registration_tree(save=True)
+        for file in reg.files.all():
+            if not dry_run:
+                file.delete()
+
+
+def purge_deleted_failed_registration_files(dry_run=False):
+    deleted_file_ids_on_failed_registration = Registration.find_failed_registrations(include_deleted=True).filter(
+        files__deleted__lte=timezone.now() - PURGE_DELTA
+    ).values_list('files__id', flat=True).distinct()
+
+    creds = Credentials.from_service_account_file(GCS_CREDS)
+    client = Client(credentials=creds)
+    for file in TrashedFileNode.objects.filter(id__in=deleted_file_ids_on_failed_registration):
+        if not dry_run:
+            file._purge(client=client)
```

Comment:
I **think** you can just wrap this in a `with transaction.atomic()` and error out of it for a dry run.